feat(Governance): design document (#213)
axelarnetwork/axelar-cgp-solidity: contracts/governance/DESIGN.md

# Axelar Governance Protocol

## Multisig

> Central to the Axelar Governance Protocol is a Multisig contract, incorporating MultisigBase and Multisig components. This contract maintains a list of signer accounts and enforces a threshold or minimum number of signers required to execute certain functions. It plays a pivotal role in defining token mint limits at the gateway.

Reviewer: ```suggestion
Central to the Axelar Governance Protocol is a Multisig contract, incorporating MultisigBase and Multisig components. This contract maintains a list of signer accounts and enforces a threshold or minimum number of signers required to execute certain functions. It plays a pivotal role in defining token mint limits for the gateway.
```

> Central to the Axelar Governance Protocol is a Multisig contract, incorporating MultisigBase and Multisig components. This contract maintains a list of signer accounts and enforces a threshold or minimum number of signers required to execute certain functions. It plays a pivotal role in defining token mint limits for the gateway.

Reviewer 2: This isn't central anymore. Main contract relies on GMP via timelocks. Multisig is only relevant for the token mint limit feature now

> 

Reviewer: ```suggestion
It plays a pivotal role in defining token mint limits for the gateway.
```

### Voting Procedure
For each transaction, a voting process is initiated among the signers. Each time a signer casts a vote, the count increments and the act is documented against the signer's address. Upon reaching the threshold, the transaction executes and the voting counter resets.

### Rotation of Signers
The Multisig contract facilitates a rotation mechanism for signers. Existing signers can refresh the list of signatory addresses and the threshold value using the `rotateSigners` function. Each rotation is tracked by incrementing an epoch.

### Execution of External Contracts

> The Multisig contract extends its purview to external contracts. The `execute` function allows signers to call any external contract. The data for this call encodes the target contract function and arguments, and may include native tokens to accompany the call.

Reviewer: ```suggestion
The Multisig contract extends its purview functionality to include external contract calls. The `execute` function allows signers to call any function on any external contract. The data accompanying this call contains the target contract's function and its associated arguments. It may also include native tokens as part of the transaction.
```

### Safeguards

> Multisig contract incorporates safeguards to deter misuse. Checks to prevent multiple votes from the same account within an epoch or duplicating a signer, as well as a requirement for transactions to meet the voting threshold before execution are in place as preventive measures.

Reviewer: ```suggestion
The Multisig contract incorporates safeguards to deter misuse. These safeguards enforce preventive measures such as prohibiting multiple votes from the same account within an epoch, preventing signer duplication, and requiring transactions to meet the voting threshold before execution.
```

## Interchain Governance

> The Interchain Governance Contract is u enables cross-chain governance actions on the Axelar network, allowing creation, cancellation, and execution of governance proposals. It institutes a TimeLock mechanism which introduces a mandatory delay before execution, safeguarding against abrupt changes.

Reviewer: ```suggestion
The Interchain Governance Contract facilitates cross-chain governance actions within the Axelar network. It enables the creation, cancellation, and execution of governance proposals while incorporating a TimeLock mechanism. The TimeLock mechanism institutes a mandatory time delay before any proposal execution, thereby offering robust protection against sudden, potentially disruptive changes.
```

### Timelock Mechanism

> The contract establishes a TimeLock mechanism ensuring secure function execution after a designated time delay. This mechanism supports scheduling, cancellation, and finalizing functions and imposes a minimum delay before any function can be scheduled or finalized.

Reviewer: ```suggestion
The TimeLock contract institutes a mechanism that guarantees the secure execution of functions after a designated time delay. This mechanism not only enables the scheduling, cancellation, and finalization of function calls, but it also enforces a minimum time delay before any function can be either scheduled or finalized, thereby enhancing the contract's security.
```

### Governance Operations
The contract manages two types of governance operations: Proposal Scheduling and Proposal Cancellation.

> - **Proposal Scheduling**: Facilitates the creation of a new proposal. With every scheduled proposal, a corresponding TimeLock is set, which must expire prior to its execution.

Reviewer: ```suggestion
- **Proposal Scheduling**:  Facilitates the creation of new proposals. For each scheduled proposal, it sets a corresponding TimeLock that must expire before the proposal's execution.
```

> - **Proposal Cancellation**: Allows the cancellation of an existing proposal, setting its corresponding TimeLock timestamp to zero and blocking its execution.

Reviewer: ```suggestion
- **Proposal Cancellation**: Facilitates the cancellation of an existing proposal by setting its corresponding TimeLock timestamp to zero and thereby blocking its execution.
```

> Both operations demand matching source chain and source address with governance chain and governance address in the contract, thereby ensuring only authorized entities have scheduling or cancellation privileges.

Reviewer: ```suggestion
Both operations require a match between the source chain and source address, and the governance chain and governance address in the contract. This check guarantees that only authorized entities can schedule or cancel actions.
```

### Secure Execution of Proposals

> Proposals can be executed once their respective TimeLock has expired. To guarantee safe execution, the contract rechecks the TimeLock immediately prior to proposal execution.

Reviewer: ```suggestion
Once their corresponding TimeLock expires, proposals become executable. For ensured safety during execution, the contract revalidates the TimeLock right before initiating the proposal execution.
```

### Generating Proposal Hashes
Uniqueness of each proposal is ensured through generating hashes employing the Keccak256 hashing algorithm on the target contract's address, function call data encoding, and the native token transfer value.

## Service Governance

> Building upon the Interchain Governance, the Service Governance Contract specializes in addressing operations requiring coordination. Incorporating a MultisigBase, it introduces functionalities to approve, execute, and cancel multisig proposals, alongside scheduling and cancelling timelock proposals.

Reviewer: ```suggestion
Building upon the Interchain Governance Contract, the Service Governance Contract is specifically designed to manage operations that require coordination. By incorporating a MultisigBase, it introduces the functionality to approve, execute, and cancel multisig proposals, as well as to schedule and cancel TimeLock proposals.
```

### Multisig Operations
Multisig operations demand multi-signatory authorization for proposal execution. A mapping mechanism tracks approvals for different proposals, requiring a proposal to garner sufficient approval to meet the predefined threshold before its execution.

### Service Governance Operations
The contract orchestrates four governance operations:

- **Schedule TimeLock Proposal**: Similar to Interchain Governance, it schedules a new governance proposal with a timelock.

> - **Cancel TimeLock Proposal**: Again, akin to Interchain Governance, it cancels an existing governance proposal.

Reviewer: ```suggestion
- **Cancel TimeLock Proposal**: Again, similar to Interchain Governance, it cancels an existing governance proposal.
```

- **Approve Multisig Proposal**: Enables multisig proposal approval, setting the approval status of the proposal to true and signaling successful approval via a `MultisigApproved` event.

> - **Cancel Multisig Approval**: Cancels an approved multisig proposal, setting the proposal’s status to false and indicating successful operation through a `MultisigCancelled` event.

Reviewer: ```suggestion
- **Cancel Multisig Approval**: Cancels an approved multisig proposal, setting the approval status of the proposal to false and indicating successful cancellation through a `MultisigCancelled` event.
```

### Secure Execution of Multisig Proposals

> A multisig proposal, upon requisite signatory approval, is eligible for execution. Pre-execution, the proposal's approval status is verified by the contract; lack of approval results in transaction reversion. Upon successful execution, the proposal’s approval status resets, triggering a `MultisigExecuted` event.

Reviewer: ```suggestion
Upon receiving the necessary number of signatory approvals, a multisig proposal becomes eligible for execution. Before execution, the contract verifies the proposal's approval status; if the approval status is false, the transaction is reverted. Following successful execution, the proposal's approval status is reset, and a `MultisigExecuted` event is emitted.
```
